x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java

```diff
@@ -132,6 +135,50 @@ private void getRoles(User user, Authentication authentication, ActionListener<R
         rolesStore.getRoles(user, authentication, listener);
     }
 
+    @Override
+    public boolean isChildActionAuthorizedByParent(RequestInfo requestInfo, String parentAction, IndicesAccessControl parentAccessControl) {
+        final String childAction = requestInfo.getAction();
+        if (IndexPrivilege.CREATE_INDEX_MATCHER.test(childAction) || childAction.equals(TransportShardBulkAction.ACTION_NAME)) {
+            // These need special handling, so don't short circuit
+            return false;
+        }
+        if (isScrollRelatedAction(childAction) || isAsyncRelatedAction(childAction)) {
+            // These need special handling, so don't short circuit
+            return false;
+        }
+        if (childAction.startsWith(parentAction) == false) {
+            // Parent action is not a true parent
+            // We want to treat shard level actions (those that append '[s]' and/or '[p]' & '[r]')
+            // or similar (e.g. search phases) as children, but not every action that is triggered
+            // within another action should be authorized this way
+            return false;
+        }
+        final IndicesRequest indicesRequest;
+        if (requestInfo.getRequest() instanceof IndicesRequest) {
+            indicesRequest = (IndicesRequest) requestInfo.getRequest();
+        } else {
+            // Can only handle indices request here
+            return false;
+        }
+
+        final String[] indices = indicesRequest.indices();
+        if (indices == null || indices.length == 0) {
+            // No indices to check
+            return false;
+        }
+
+
+        return Arrays.stream(indices).allMatch(idx -> {
+            if (Regex.isSimpleMatchPattern(idx)) {
+                // The request contains a wildcard
+                return false;
+            }
```

Comment:
This check means the performance boost does not apply to shard level requests if the original request has wildcard since `indices()` returns the originalIndices. I think this is what we wanted for this PR to solve a particular problem. But it's worthwhile to build upon it and cover more shard level requests in future.

Reply:
I don't think that's true. The wildcards are expanded and replaced in the original request, so all the children have expanded sets as well.
More pragmatically, this PR definitely improves the performance of `/*/_search` so it must work :)

Reply:
You are right. The wildcards are expanded/resolved when authorizing for the first time. The resolved indices replace the original indices with `IndicesRequest#Replacable#indices` (we effectively drop the original indices along the way). Hence child actions only see the resolved indices.